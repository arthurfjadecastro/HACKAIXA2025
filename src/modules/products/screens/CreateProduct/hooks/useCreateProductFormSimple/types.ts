import { ScrollView } from 'react-native';
import { FormData, FieldName } from '../../types';

export interface UseSimpleFormReturn {
  formData: FormData;
  isLoading: boolean;
  isFormValid: boolean;
  updateField: (fieldName: FieldName, value: string | number | boolean | string[]) => void;
  handleSubmit: () => Promise<void>;
  scrollViewRef: React.RefObject<ScrollView | null>;
  isConvenioAlreadyRegistered: (convenioKey: string) => boolean;
  isHabitacaoAlreadyRegistered: () => boolean;
}

export interface ConvenioMatches {
  [key: string]: string[];
  inss: string[];
  militar: string[];
  funcef: string[];
  tjdft: string[];
}

export const CONVENIO_MATCHES: ConvenioMatches = {
  'inss': ['inss', 'aposentados', 'pensionistas'],
  'militar': ['militar', 'aeronáutica', 'comando'],
  'funcef': ['funcef'],
  'tjdft': ['tjdft', 'tribunal']
};

export const INITIAL_FORM_DATA: FormData = {
  categoria: '',
  subtipo: '',
  name: '',
  interestRate: '',
  maxTerm: '',
  normative: '',
  observacoes: [],
  fonte_dados: 'Cadastro automático',
  // Valores padrão para campos de OUTRO
  prazo_min_meses: 1,
  prazo_max_meses: 420
};
