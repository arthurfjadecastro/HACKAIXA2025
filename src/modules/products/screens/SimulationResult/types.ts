import { RouteProp } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { AppStackParamList } from '@/navigation/AppStack';
import { LoanCalculationResult } from '@/utils/loanCalculator';

export type RouteProps = RouteProp<AppStackParamList, 'SimulationResult'>;
export type NavigationProps = NativeStackNavigationProp<AppStackParamList>;

export interface SimulationParams {
  productId: string;
  amount: string;
  months: number;
  result: {
    rate: number;
    rateAnnual?: number;
    amortizationType?: 'PRICE' | 'SAC';
  };
}

// Re-export the existing type from loanCalculator
export type LoanSchedule = LoanCalculationResult;
