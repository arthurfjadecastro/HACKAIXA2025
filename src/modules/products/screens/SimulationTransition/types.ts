import { NativeStackNavigationProp } from '@react-navigation/native-stack';

export type NavigationProps = NativeStackNavigationProp<any>;

export type SimulationState = 'idle' | 'loading' | 'success' | 'error';

export interface SimulationResult {
  installment: number;
  total: number;
  rate: number;
  months: number;
  amount: number;
}

export interface SimulationParams {
  productId: string;
  amount: string;
  months: number;
}

export const STATUS_TEXTS = [
  'Validando dados…',
  'Calculando parcelas…',
  'Confirmando cenário…'
] as const;

export const MIN_DWELL_TIME = 900; // 900ms mínimo
export const TIMEOUT_DURATION = 15000; // 15 segundos
export const STATUS_ROTATION_INTERVAL = 1200; // 1.2 segundos
