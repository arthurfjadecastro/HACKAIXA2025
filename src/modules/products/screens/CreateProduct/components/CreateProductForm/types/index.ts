import { FormData, FieldName } from '../../../types';

export interface CreateProductFormProps {
  formData: FormData;
  updateField: (fieldName: FieldName, value: string | number | boolean | string[]) => void;
  isConvenioAlreadyRegistered: (convenioKey: string) => boolean;
  isHabitacaoAlreadyRegistered: () => boolean;
}

export interface CategoryOption {
  value: string;
  label: string;
  description: string;
  disabled?: boolean;
}
