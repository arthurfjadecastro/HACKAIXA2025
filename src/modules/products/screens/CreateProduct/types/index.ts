export interface FormData {
  // Seleção de categoria/subtipo
  categoria: 'CONSIGNADO' | 'HABITACAO' | 'OUTRO' | '';
  subtipo: 'INSS' | 'CONVENIO' | 'CLT_SUSPENSO' | 'N/A' | '';
  convenio_id?: string;
  convenio_selected?: string; // Para seleção simplificada (militar, funcef, tjdft)
  
  // Campos originais
  name: string;
  interestRate: string;
  maxTerm: string;
  normative: string;
  
  // Campos preenchidos automaticamente - Consignado
  prazo_minimo?: number;
  prazo_maximo?: number;
  margem_consignavel?: number;
  canais_permitidos?: string[];
  taxa_faixa_a_concessao?: number;
  taxa_faixa_a_renovacao?: number;
  taxa_faixa_b_concessao?: number;
  taxa_faixa_b_renovacao?: number;
  taxa_faixa_c_concessao?: number;
  taxa_faixa_c_renovacao?: number;
  portabilidade_prospeccao_min?: number;
  portabilidade_prospeccao_max?: number;
  exige_validacao_remota?: boolean;
  exige_validacao_presencial?: boolean;
  exige_certificacao_digital?: boolean;
  
  // Campos específicos de Habitação
  sistema_amortizacao?: 'SAC' | 'PRICE' | '';
  ltv_max_percentual?: number;
  entrada_min_percentual?: number;
  indexadores_permitidos?: Array<{id: string, descricao: string}>;
  seguros_obrigatorios?: Array<{id: string, descricao: string}>;
  
  // Campos específicos de OUTRO
  subcategoria_outro?: string;
  valor_min?: number;
  valor_max?: number;
  prazo_min_meses?: number;
  prazo_max_meses?: number;
  quantidade_min?: number;
  quantidade_max?: number;
  taxa_percentual?: number;
  taxa_valor_fixo?: number;
  taxa_por_unidade?: number;
  taxa_observacao?: string;
  canais_selecionados?: string[];
  faixa_etaria_min?: number;
  faixa_etaria_max?: number;
  
  // Campos comuns
  observacoes?: string[];
  fonte_dados?: string;
}

export interface FormErrors {
  categoria?: string;
  subtipo?: string;
  convenio_id?: string;
  name?: string;
  interestRate?: string;
  maxTerm?: string;
  normative?: string;
}

export interface FieldValidation {
  isValid: boolean;
  error?: string;
}

export type FieldName = keyof FormData;

export interface UseFormReturn {
  formData: FormData;
  errors: FormErrors;
  touched: Record<string, boolean>;
  isLoading: boolean;
  isFormValid: boolean;
  updateField: (fieldName: FieldName, value: string | number | boolean | string[]) => void;
  handleBlur: (fieldName: FieldName) => void;
  handleSubmit: () => Promise<void>;
  scrollToFirstError: () => void;
}
