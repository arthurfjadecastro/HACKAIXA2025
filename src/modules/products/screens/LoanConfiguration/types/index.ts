export interface QuoteData {
  installment: number;
  total: number;
  rate: number;
}

export interface ProductData {
  maxMonths: number;
  minMonths: number;
  rateAm: number;
}

export interface ValidationState {
  showWarning: boolean;
  timeout: NodeJS.Timeout | null;
}

export interface LoanConfigurationProps {
  productId: string;
  amount: string;
}

export interface InstallmentSelectorProps {
  months: number;
  inputValue: string;
  productMinMonths: number;
  productMaxMonths: number;
  showValidationWarning: boolean;
  onMonthsChange: (value: string) => void;
  onDecrease: () => void;
  onIncrease: () => void;
}

export interface AmountSummaryProps {
  amount: string;
  onEdit: () => void;
}

export interface InstallmentValueProps {
  quote: QuoteData | null;
}

export interface FinancialSummaryProps {
  quote: QuoteData | null;
  productRateAm: number;
}

export interface LoanHeaderProps {
  onBackPress: () => void;
}
