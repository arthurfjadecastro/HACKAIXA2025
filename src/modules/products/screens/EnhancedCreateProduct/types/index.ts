import { ProductFormData, ProductCategory } from '@/types/products';

export interface EnhancedCreateProductProps {
  // Props futuras se necess√°rio
}

export interface FormState {
  formData: ProductFormData;
  selectedConvenio: ProductCategory | null;
  loading: boolean;
}

export interface CategoryOption {
  value: string;
  label: string;
  description: string;
}
