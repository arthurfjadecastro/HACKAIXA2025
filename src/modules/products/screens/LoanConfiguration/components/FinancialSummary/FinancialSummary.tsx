import React from 'react';
import { View, Text } from 'react-native';
import { styles } from '../../LoanConfiguration.styles';
import { QuoteData } from '../../types';

interface FinancialSummaryProps {
  quote: QuoteData | null;
  productRateAm: number;
  formatCurrency: (value: number) => string;
}

export const FinancialSummary: React.FC<FinancialSummaryProps> = ({
  quote,
  productRateAm,
  formatCurrency,
}) => {
  return (
    <>
      {/* Valor por parcela */}
      <View style={styles.installmentValue}>
        <Text style={styles.installmentLabel}>de</Text>
        <Text style={styles.installmentAmount}>
          {quote ? formatCurrency(quote.installment) : 'R$ 0,00'}
        </Text>
      </View>

      {/* Resumo financeiro */}
      <View style={styles.financialSummary}>
        <View style={styles.summaryRow}>
          <Text style={styles.summaryLabel}>Total a pagar</Text>
          <Text style={styles.summaryValue}>
            {quote ? formatCurrency(quote.total) : 'R$ 0,00'}
          </Text>
        </View>
        
        <View style={styles.summaryRow}>
          <Text style={styles.summaryLabel}>Juros</Text>
          <Text style={styles.summaryValue}>
            {(productRateAm * 100).toFixed(2)}% a.m. ({((Math.pow(1 + productRateAm, 12) - 1) * 100).toFixed(1)}% a.a. efetiva)
          </Text>
        </View>
      </View>
    </>
  );
};
