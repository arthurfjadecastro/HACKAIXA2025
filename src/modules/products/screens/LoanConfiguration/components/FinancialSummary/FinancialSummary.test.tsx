import React from 'react';
import { render } from '@testing-library/react-native';
import { FinancialSummary } from './FinancialSummary';
import { QuoteData } from '../../types';

// Mock dos styles
jest.mock('../../LoanConfiguration.styles', () => ({
  styles: {
    installmentValue: { padding: 10 },
    installmentLabel: { fontSize: 14 },
    installmentAmount: { fontSize: 24 },
    financialSummary: { padding: 15 },
    summaryRow: { flexDirection: 'row' },
    summaryLabel: { fontSize: 16 },
    summaryValue: { fontSize: 16 },
  },
}));

describe('FinancialSummary', () => {
  const mockQuote: QuoteData = {
    installment: 2500,
    total: 60000,
    rate: 0.025,
  };

  const defaultProps = {
    quote: mockQuote,
    productRateAm: 0.025,
    formatCurrency: jest.fn((value: number) => `R$ ${value.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with quote data', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} />);

    expect(getByText('de')).toBeTruthy();
    expect(getByText('Total a pagar')).toBeTruthy();
    expect(getByText('Juros')).toBeTruthy();
  });

  it('should display installment amount when quote is provided', () => {
    render(<FinancialSummary {...defaultProps} />);

    expect(defaultProps.formatCurrency).toHaveBeenCalledWith(2500);
  });

  it('should display total amount when quote is provided', () => {
    render(<FinancialSummary {...defaultProps} />);

    expect(defaultProps.formatCurrency).toHaveBeenCalledWith(60000);
  });

  it('should display default values when quote is null', () => {
    const { getAllByText } = render(<FinancialSummary {...defaultProps} quote={null} />);

    // Verifica se existe pelo menos um elemento com R$ 0,00
    const zeroValues = getAllByText('R$ 0,00');
    expect(zeroValues.length).toBeGreaterThan(0);
  });

  it('should calculate and display interest rates correctly', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} />);

    // Monthly rate: 2.50% a.m.
    expect(getByText('2.50% a.m. (34.5% a.a. efetiva)')).toBeTruthy();
  });

  it('should handle different interest rates', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} productRateAm={0.03} />);

    // Monthly rate: 3.00% a.m.
    expect(getByText('3.00% a.m. (42.6% a.a. efetiva)')).toBeTruthy();
  });

  it('should handle zero interest rate', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} productRateAm={0} />);

    expect(getByText('0.00% a.m. (0.0% a.a. efetiva)')).toBeTruthy();
  });

  it('should match snapshot with quote', () => {
    const component = render(<FinancialSummary {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot without quote', () => {
    const component = render(<FinancialSummary {...defaultProps} quote={null} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render installment label', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} />);

    expect(getByText('de')).toBeTruthy();
  });

  it('should handle high interest rates correctly', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} productRateAm={0.1} />);

    expect(getByText('10.00% a.m. (213.8% a.a. efetiva)')).toBeTruthy();
  });

  it('should display all summary sections', () => {
    const { getByText } = render(<FinancialSummary {...defaultProps} />);

    expect(getByText('de')).toBeTruthy();
    expect(getByText('Total a pagar')).toBeTruthy();
    expect(getByText('Juros')).toBeTruthy();
  });
});
