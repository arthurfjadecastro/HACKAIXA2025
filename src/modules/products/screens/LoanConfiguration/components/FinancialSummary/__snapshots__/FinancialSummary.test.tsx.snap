// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinancialSummary should match snapshot with quote 1`] = `
[
  <View
    style={
      {
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 14,
        }
      }
    >
      de
    </Text>
    <Text
      style={
        {
          "fontSize": 24,
        }
      }
    >
      R$ 2.500,00
    </Text>
  </View>,
  <View
    style={
      {
        "padding": 15,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        Total a pagar
      </Text>
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        R$ 60.000,00
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        Juros
      </Text>
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        2.50
        % a.m. (
        34.5
        % a.a. efetiva)
      </Text>
    </View>
  </View>,
]
`;

exports[`FinancialSummary should match snapshot without quote 1`] = `
[
  <View
    style={
      {
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 14,
        }
      }
    >
      de
    </Text>
    <Text
      style={
        {
          "fontSize": 24,
        }
      }
    >
      R$ 0,00
    </Text>
  </View>,
  <View
    style={
      {
        "padding": 15,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        Total a pagar
      </Text>
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        R$ 0,00
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        Juros
      </Text>
      <Text
        style={
          {
            "fontSize": 16,
          }
        }
      >
        2.50
        % a.m. (
        34.5
        % a.a. efetiva)
      </Text>
    </View>
  </View>,
]
`;
